[Dumper] Add support for XmlReader objets

Taluu opened the pull request:
| Q | A |
| --- | --- |
| Branch? | "master" |
| Bug fix? | no |
| New feature? | yes |
| BC breaks? | no |
| Deprecations? | no |
| Tests pass? | yes |
| Fixed tickets | ~ |
| License | MIT |
| Doc PR | ~ |

Adding support for `XmlReader` objects. The thing is, with the caster as presented here, we may have noises (`#text` elements, whitespaces (significant or not) elements, ...) which may disturb the reading, but I am not sure if I should ignore those or not.

Moreover, there is only one class, but having different casters could do the trick I guess ? But then we have a lot of noise with empty results if none of the casters are satisfied (unless I missed something, allowing us to completely ignore the current object, in cases such as `XmlReader::WHITESPACE` and `XmlReader::SIGNIFICANT_WHITESPACE`). 

On attributes, I can either not touch the current state, but then I can only have indexes for attributes and no names (`<foo bar="baz">` would be rendered as `array(0 => 'baz')`), or I can modify the object (rendering `<foo bar="baz">` as `array('bar' => 'baz')`) , but as @stof said, we should not touch it, unless it is not the actual object that is modified ? I do not have enough knowledge on the var dumper to determine that point...

Open for suggestions.
- [x] Add basic XmlReaderCaster
- [x] Add XmlReaderCaster in the default casters
- [x] Add some tests

Taluu committed:
Add support for XmlReader objets

Taluu committed:
Add XmlReaderCaster in the default casters

stof commented on src/Symfony/Component/VarDumper/Caster/XmlReaderCaster.php:
```diff
@@ -0,0 +1,73 @@
+class XmlReaderCaster
+{
+    private static $nodeTypes = array(
+            XmlReader::NONE => 'NONE',
```
wrong indentation

Taluu replied:
oops, fixed

nicolas-grekas replied:
& missing `\` (XmlReader is not in `Symfony\Component\VarDumper\Caster`)

stof replied:
@Taluu I suggest you to start writing tests :smile:

Taluu commented:
Note ; I'm also not sure if I should dump _all_ the xml, or just dump the current step (as it is done right now)

stof commented on src/Symfony/Component/VarDumper/Caster/XmlReaderCaster.php:
```diff
@@ -0,0 +1,73 @@
+        if ($reader->hasAttributes) {
+            $infos['attributes'] = array();
+
+            while ($reader->moveToNextAttribute()) {
```
this changes the state of the reader, which is a no-go. The casters are not allowed to produce side effects.

Taluu replied:
Yes, I had a different implementation before, but I can't have the attributes name (only indexes). If that is enough, then I guess I could revert back.

But aren't the variables cloned or something like that when passed to the casters ? /cc @nicolas-grekas

nicolas-grekas replied:
This changes the state of the $reader object, is it possible to get the attributes doing any "move"?

Taluu replied:
Using something like this (Which is what I did before, but thought it was not really generous in information) :

```php
<?php
for ($i = 0; $c = $reader->attributeCount; $i < $c; ++$i) {
    $infos[Caster::PREFIX_VIRTUAL . 'attributes'][] = $reader->getAttributeNo($i);
}
```

But then we are loosing the attribute name...

stof commented:
Tests should be added

nicolas-grekas commented on src/Symfony/Component/VarDumper/Caster/XmlReaderCaster.php:
```diff
@@ -0,0 +1,73 @@
+    {
+        $infos = array(
+            'name' => $reader->name,
+            'local_name' => $reader->localName,
```
you should stick to the public property name (`localName`) so that one can learn by example how to use the object

Taluu replied:
fixed. I am then exposing `'localName' => $reader->localName`, to stick with your other comments (and use the real property name)

nicolas-grekas commented on src/Symfony/Component/VarDumper/Caster/XmlReaderCaster.php:
```diff
@@ -0,0 +1,73 @@
+
+            'depth' => $reader->depth,
+
+            'attributes_count' => $reader->attributeCount,
```
attributeCount

Taluu replied:
fixed

nicolas-grekas commented on src/Symfony/Component/VarDumper/Caster/XmlReaderCaster.php:
```diff
@@ -0,0 +1,73 @@
+
+            'has_value' => $reader->hasValue,
+            'is_default' => $reader->isDefault,
+            'is_empty' => $reader->isEmptyElement,
```
isEmptyElement

Taluu replied:
fixed

nicolas-grekas commented on src/Symfony/Component/VarDumper/Caster/XmlReaderCaster.php:
```diff
@@ -0,0 +1,73 @@
+        }
+
+        if ($reader->hasAttributes) {
+            $infos['attributes'] = array();
```
this is a virtual property (`$reader->attributes` does not exist isn't it?)
thus it should be `$infos[Caster::PREFIX_VIRTUAL.'attributes']`

Taluu replied:
fixed

Taluu committed:
Do not touch XmlReader state while reading attributes

stof commented on src/Symfony/Component/VarDumper/Caster/XmlReaderCaster.php:
```diff
@@ -0,0 +1,72 @@
+        \XmlReader::SIGNIFICANT_WHITESPACE => 'SIGNIFICANT_WHITESPACE',
+        \XmlReader::END_ELEMENT => 'END_ELEMENT',
+        \XmlReader::END_ENTITY => 'END_ENTITY',
+        \XmlReader::XML_DECLARATION => 'XML_DECLARATION'
```
This does not contain all possible types

Taluu committed:
Ignore (partially) some node types

Taluu commented:
Hum, just thought of something else for the attributes, and getting the name. How about moving to the attribute node (effectively modifying the object) with [`moveToAttributeNo`](http://nl1.php.net/manual/en/xmlreader.movetoattributeno.php), but then reverting back to the current element with [`moveToElement`](http://nl1.php.net/manual/en/xmlreader.movetoelement.php) ?

I could then get the name and the value of the attribute. But it does imply a little moving and going backwards though (what if there is an error in between ?).

What do you think of this @stof, @nicolas-grekas ? With a big fat warning if needed so ?

stof commented:
> With a big fat warning if needed so ?

This won't help much, as this warning would only be seen by developers of the component having to make it work, which means it should just work from the start (your big fat warning would be nothing more than a FIXME comment in practice).

nicolas-grekas commented:
Maybe use `readOuterXML` or `expand`? I agree with stof, the warning would be useless.

Taluu commented:
One of the point of using `XMLReader` is to be able to load a huge file, as DOM or SimpleXML needs to load everything in one go, then making the loading go berserk.... So using expand (which transforms the current node in DOM) or translating it as `outerXML` would I think defeat one of the point of using XMLReader.

And using that just to get the names of the attributes looks a bit overkill to me....

Taluu commented:
After I gave it some thoughts, I think that using `expand` would not be really overkill (we don't want to dump ALL the nodes, just one node at a given time.. ?). Then exposing the `expand` as a virtual method could be useful too (and using this to determine the attributes too of course) ?

Taluu commented:
ping @stof, @nicolas-grekas ; I am now managing the attributes with `expand()`, and I even exposed the dom as a cut stub (and virtual prefix).

I am not sure the dom should not be cut, but imo it kind of duplicates the main information of what is given by the main caster.

WDYT ? I'll add tests if that's good. 

Output example : https://gist.github.com/Taluu/24c6c908fda299069c756a2be38f81dc

nicolas-grekas commented on src/Symfony/Component/VarDumper/Caster/XmlReaderCaster.php:
```diff
@@ -0,0 +1,102 @@
+            'isEmptyElement' => $reader->isEmptyElement,
+            'nodeType' => new ConstStub(self::$nodeTypes[$reader->nodeType], $reader->nodeType),
+
+            Caster::PREFIX_VIRTUAL.'dom' => new CutStub($dom)
```
I'd suggest calling it expand, to help discoverability

Taluu replied:
renamed into `expand`

nicolas-grekas commented on src/Symfony/Component/VarDumper/Caster/XmlReaderCaster.php:
```diff
@@ -0,0 +1,102 @@
+
+        if (false === static::$filteredTypes[$reader->nodeType]) {
+            $stub->cut += count($infos);
+            return array();
```
why that? can't we at least keep nodeType & maybe a few relevant others?

Taluu replied:
Added `localName`, `depth` and `nodeType`. I don't think the other are really useful though...

nicolas-grekas commented on src/Symfony/Component/VarDumper/Caster/XmlReaderCaster.php:
```diff
@@ -0,0 +1,102 @@
+    protected static $filteredTypes = array(
+        \XmlReader::NONE => true,
+        \XmlReader::ELEMENT => true,
+        \XmlReader::ATTRIBUTE => false,
```
I suggest keeping only filttered type here (remove all "false" items) and use an isset below

Taluu replied:
Thought it would be better to have all the types so we can easily mark some more as filtered, but adding a new entry is easy too. Fixed.

nicolas-grekas commented:
some comments posted, time to add tests to me (and fix fabbot issue) :)

Taluu committed:
Add unit tests for XmlReader Caster

Taluu commented:
Tests added /cc @nicolas-grekas, @stof

stof commented on src/Symfony/Component/VarDumper/Caster/XmlReaderCaster.php:
```diff
@@ -0,0 +1,112 @@
+
+    public static function castXmlReader(\XmlReader $reader, array $a, Stub $stub, $isNested)
+    {
+        $dom = $reader->expand();
```
Wouldn't this convert the whole XML file into a DOM object in case the node being dumped is the root node ? This would make dumping incompatible with huge XML files

Taluu replied:
Yes, this was my point in an earlier comment, but I am not sure one would really dump such a huge xml with this... ?

stof replied:
if you are dumping the reader while debugging your code, you might not know which node is currently in the reader (otherwise you would not need to dump it)

Taluu replied:
True, but then I think I do not have any solutions for properly dumping attributes (or the `expand()` result actually)

stof replied:
I don't think dumping the ``expand`` result is needed

nicolas-grekas replied:
the difficulty comes from reading attributes, let's not dump them!

Taluu replied:
But then, I can dump them, but without their name (only their values). `getAttributeNo` does not consume the reader.

nicolas-grekas commented on src/Symfony/Component/VarDumper/Caster/XmlReaderCaster.php:
```diff
@@ -0,0 +1,112 @@
+            'hasValue' => $reader->hasValue,
+            'isDefault' => $reader->isDefault,
+            'isEmptyElement' => $reader->isEmptyElement,
+            'nodeType' => $nodeType,
```
btw looking at the doc, there are more properties: http://php.net/manual/en/class.xmlreader.php
we should even dump results of `getParserProperty` IMHO

Taluu replied:
I added the prefix and the namespaceURI if the prefix is not empty (had to check `!== ''`, because even if it has a value (such as `baz`), `empty($reader->prefix)` returns `true`...)

And I also added the parser parameters, even though I am not really convinced. I also added it when the elements are ignored.

Taluu committed:
Add some more info on ATTRIBUTES elements

Taluu committed:
Add parser properties virtual property

Taluu committed:
Add prefix and namespaceURI if declared

Taluu committed:
Add special case for NONE node types

nicolas-grekas commented on src/Symfony/Component/VarDumper/Caster/XmlReaderCaster.php:
```diff
@@ -0,0 +1,141 @@
+        $infos = array(
+            'nodeType' => $nodeType,
+            Caster::PREFIX_VIRTUAL.'parserProperties' => array(
+                'LOADDTD' => $reader->getParserProperty(\XmlReader::LOADDTD),
```
these should be wrapped as ConstStub

stof replied:
why a const stub ? parser properties are just booleans

nicolas-grekas replied:
true, looks like I meant EnumStub on the array as already suggested :)

nicolas-grekas commented on src/Symfony/Component/VarDumper/Caster/XmlReaderCaster.php:
```diff
@@ -0,0 +1,141 @@
+            'isEmptyElement' => $reader->isEmptyElement,
+        );
+
+        if ($reader->prefix !== '') {
```
`'' !== $reader->prefix)

nicolas-grekas commented on src/Symfony/Component/VarDumper/Caster/XmlReaderCaster.php:
```diff
@@ -0,0 +1,141 @@
+
+        $infos = array(
+            'nodeType' => $nodeType,
+            Caster::PREFIX_VIRTUAL.'parserProperties' => array(
```
` => new EnumStub(array(`...
maybe also put this item always last is the list?

Taluu replied:
As you see, I have some early returns, and I thought it would always be useful. But I can still use a `try {} finally {}` block and add these at the end... ?

stof replied:
don't use try/finally for that. Using return statements inside finally makes things very hard to understand due to their weird behavior

Taluu replied:
I agree, but then I have to repeat myself in order to place said block at the end of the list

Anyhow, I had plans to refactor and split this method because it is becoming kind of huge with too many paths

stof commented on src/Symfony/Component/VarDumper/Tests/Caster/XmlReaderCasterTest.php:
```diff
@@ -0,0 +1,318 @@
+
+namespace Symfony\Component\VarDumper\Tests\Caster;
+
+use XmlReader;
```
we don't add use statements for global classes

Taluu committed:
Use an Enum Stub for parser properties

Taluu committed:
Move the parserProperties at the end of list

Taluu committed:
Refactor XmlReader caster

Split main method into several sub methods when handling "special" cases
such as ATTRIBUTE, NONE or filtered elements

Taluu commented:
I split the method (the special cases into their own methods) so that it is clearer. I am not a huge fan of the "NUMBER" constant, but I didn't want to calculate the whole array just to discard it anyway at the end...

Any ideas are welcome though.

nicolas-grekas commented on src/Symfony/Component/VarDumper/Tests/Caster/XmlReaderCasterTest.php:
```diff
@@ -0,0 +1,317 @@
+
+    public function testNone()
+    {
+        $dump = <<<DUMP
```
`<<<'DUMP'`

nicolas-grekas commented on src/Symfony/Component/VarDumper/Tests/Caster/XmlReaderCasterTest.php:
```diff
@@ -0,0 +1,317 @@
+
+    public function testPrefixedElements()
+    {
+        $dump = <<<DUMP
```
`<<<'DUMP'`

nicolas-grekas commented on src/Symfony/Component/VarDumper/Tests/Caster/XmlReaderCasterTest.php:
```diff
@@ -0,0 +1,317 @@
+
+    public function testAttributeNode()
+    {
+        $dump = <<<DUMP
```
`<<<'DUMP'`

nicolas-grekas commented on src/Symfony/Component/VarDumper/Caster/XmlReaderCaster.php:
```diff
@@ -0,0 +1,195 @@
+        \XmlReader::END_ENTITY => true,
+    );
+
+    /** @const int Number of elements in the stub, for all elements */
```
should be marked `@internal` or let's inline `9`, a comment is enough

stof replied:
thus, inlining it will avoid having to add 9 to the cut number, and then to substract another value due to cutting it only partially. We will be able to inline the ``6`` directly for attributes for instance

nicolas-grekas commented on src/Symfony/Component/VarDumper/Caster/XmlReaderCaster.php:
```diff
@@ -0,0 +1,195 @@
+
+        // ignore localName if it is a #text node
+        //
+        // (No significant information brought by this name, the nodeType is
```
I don't get this comment block (no need for parenthesis btw), can't we make it more clear?

Taluu replied:
The point was that having the following :

```
XMLReader {
    localName: "#text"
    nodeType: "TEXT"
}
```

(Or `Whitespace`, ... etc) is redundant, so I chose to ignore the `localName`. But I can readd it...

nicolas-grekas replied:
My personal strategy in these situations is to KISS: being cleaver is too hard, that's how things are, let's not hide the truth.

nicolas-grekas commented on src/Symfony/Component/VarDumper/Caster/XmlReaderCaster.php:
```diff
@@ -0,0 +1,195 @@
+        $infos = array(
+            'localName' => $reader->localName,
+            'nodeType' => $nodeType,
+
```
let's drop these blank lines

stof commented on src/Symfony/Component/VarDumper/Caster/XmlReaderCaster.php:
```diff
@@ -0,0 +1,195 @@
+        \XmlReader::XML_DECLARATION => 'XML_DECLARATION',
+    );
+
+    protected static $filteredTypes = array(
```
should be private

nicolas-grekas commented on src/Symfony/Component/VarDumper/Caster/XmlReaderCaster.php:
```diff
@@ -0,0 +1,195 @@
+    /** @const int Number of elements in the stub, for all elements */
+    const BASE_ITEM_NUMBER = 9;
+
+    public static function castXmlReader(\XmlReader $reader, array $a, Stub $stub, $isNested)
```
I would adopt a different code organization: let's start by building the full $infos array without *any* conditional statement, then filter base on the type. This should help reduce duplications, and ease counting `cut` value.

Taluu replied:
Can't do that, as NONE elements triggers fatal errors when trying to access the name or other properties. I thought of making some "special" cases, such as NONE and ATTRIBUTES, but then I could treat the filter afterwards and make a difference in the number of elements, as was done before.

nicolas-grekas commented:
Continued in #18989
If you're OK @Taluu , I'll let you close this one.

Taluu commented:
Closing then
